#### 2024.9.18

**今日学习：
（67）项目实践·引导（68）项目实践·英国电商销售数据**

#### 一、项目介绍

**（1）整体流程:**

        下载数据--读取数据--评估数据--根据评估得到的结论解决发现的问题--每解决一个问题便再次查看清洗后的数据确保问题解决--保存清洗后的数据

**（2）数据集介绍:**

        这是一个跨国数据集，其中包含2010年12月12日至2011年12月9日期间发生的所有在英国注册的非商店在线零售业务的交易；
        该公司主要销售独特的全场礼品，并且大部分客户是批发商

**（3）字段含义:**

        数据形状为：542k 行x 8列，8个字段分别为发票号，发票日期，商品码，商品描述，数量，单价，顾客ID，国家
        
        
        InvoiceNo: 订单号码，每笔交易分配唯一的6位整数，而退货订单的代码以字母'c'开头
        
        StockCode: 产品代码，每个不同的产品分配唯一的5位整数
        
        Description: 产品描述，对每件产品的简略描述
        
        Quantity: 产品数量，每笔交易的每件产品的数量
        
        InvoiceDate: 订单日期和时间，每笔交易发生的日期和时间
        
        UnitPrice: 单价（英镑），单位产品价格
        
        CustomerID:顾客号码，每个客户分配唯一的5位整数
        
        Country: 国家的名字，每个客户所在国家/地区的名称

**（4）分析目标:**

        此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。
        本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

=========================================================================================================================

==============================================================================================================

#### 二、项目分析

In [1]:
import pandas as pd

## 下载数据

In [2]:
original_data = pd.read_csv("/Users/warma/Desktop/python/实践项目：英国电商销售数据/e_commerce.csv", encoding='ISO-8859-1')  
# 尝试使用 'ISO-8859-1' 编码

## 读取数据

In [3]:
original_data.head()    # 提取前五行数据

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

### 评估数据整齐度

In [4]:
original_data.sample(10)    # 随便提取十行数据

InvoiceNo StockCode                       Description  Quantity  \
481674    577358    82616B     FRAPPUCINO SCARF KNITTING KIT         1   
225657    556728    82494L       WOODEN FRAME ANTIQUE WHITE          2   
111111    545713     22499         WOODEN UNION JACK BUNTING         4   
245152    558613     22904         CALENDAR PAPER CUT DESIGN        24   
109101    545583     22139  RETROSPOT TEA SET CERAMIC 11 PC          3   
370207    569132     21494    ROTATING LEAVES T-LIGHT HOLDER         1   
16287     537646    72799F    IVORY PILLAR CANDLE GOLD FLOCK         1   
513237    579541     23557      LANDMARK FRAME BAKER STREET          1   
486408    577729    85099B           JUMBO BAG RED RETROSPOT        10   
350828    567656     23170          REGENCY TEA PLATE ROSES          1   

             InvoiceDate  UnitPrice  CustomerID         Country  
481674  11/18/2011 15:59       2.46         NaN  United Kingdom  
225657   6/14/2011 11:36       2.95     15719.0  United Kingdom  
111111    3/7/2011 10:11       8.29         NaN  United Kingdom  
245152   6/30/2011 15:46       2.95     14911.0            EIRE  
109101     3/4/2011 9:08       4.95     12553.0          France  
370207   9/30/2011 13:54       1.25     17757.0  United Kingdom  
16287    12/7/2010 15:39       2.95     17591.0  United Kingdom  
513237  11/30/2011 10:12      12.50     14653.0  United Kingdom  
486408  11/21/2011 12:53       2.08     15179.0  United Kingdom  
350828   9/21/2011 14:40       3.29     14096.0  United Kingdom

### 评估数据干净度

In [5]:
original_data.info()    # 查看数据基本情况

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


### 评估缺失数据

**（1）Description：**
        
        由 Description  540455 non-null  object 

        发现 Description 存在缺失值，便提取出 Description 的所有缺失值

In [6]:
original_data["Description"].isnull()    # 返回 "Description" 行的 Bool值

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: Description, Length: 541909, dtype: bool

In [7]:
original_data[original_data["Description"].isnull()]    # 根据 Bool值 Series 中的 True，返回 Description 的所有缺失值

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

        通过初步观察，发现有一个同一特征，即 UnitPrice 全部等于0

        但这不一定是导致缺失值的原因，仅是一个猜测，我们接下来进行验证

In [8]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

        输出结果展示每一行 "Description" 存在缺失值的数据 ，其 "UnitPrice" 数据均为0

        即缺失产品名的数据，其产品单价也是0

**结论：**

        由于我们的分析目标是挖掘畅销产品，因此上述没有单价的产品数据并不在我们分析的范围内，可以删除

**（2）CustomerID：**
        
        由 CustomerID   406829 non-null  float64

        发现 CustomerID 存在缺失值，便提取出 CustomerID 的所有缺失值

In [9]:
original_data["CustomerID"].isnull()    # 返回 "CustomerID" 行的 Bool值

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: CustomerID, Length: 541909, dtype: bool

In [10]:
original_data[original_data["CustomerID"].isnull()]    # 根据 Bool值 Series 中的 True，返回 CustomerID 的所有缺失值

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

**结论：**

        由于我们的分析目标是挖掘畅销产品，因此上述没有客户ID的产品数据并不影响我们进行研究，因此可以保留

### 评估重复数据

        唯一标识符，如：
                    InvoiceNo: 订单号码，每笔交易分配唯一的6位整数，而退货订单的代码以字母'c'开头。
        
                    StockCode: 产品代码，每个不同的产品分配唯一的5位整数

                    CustomerID:顾客号码，每个客户分配唯一的5位整数

        但一次交易也可能包含多件商品，便可能出现一个发票编号对应多个商品的观察值（如前几行）

**结论：**

        由于以上多种可能的情况，针对次数据集，我们无需评估重复数据

### 评估不一致数据

        指同一个含义有不同的表达方式，如表达中国即有China、也有china

        这要求我们对数据有一定的了解

In [11]:
original_data["Country"].value_counts()    # 提取出 "Country" 列所有值以及其各自的出现次数

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


            我们可以发现上述结果中：从Country变量值来看 "USA"、"United States"均在表示美国，
                                "United Kingdom"、"UK"、"U.K."均在表示英国 因此应该在清洗步骤对这些值进行统一，只保留一个指代值

                                其实我这份数据并没有这个问题，说明数据原本修改过了，以上是教学视频中出现的问题，再次进行提醒

### 评估无效或错误数据

            可以通过DataFrame的describe方法对数值统计信息进行快速了解

In [12]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

            从输出结果来看Quantity和UnitPrice存在负数，会对后续数值分析产生影响 
            
            因此 我们先筛选出Quantity和UnitPrice数值为负数的观察值，进一步评估其含义

In [13]:
original_data[original_data["Quantity"] < 0]    # 提取 "Quantity" 中销量为负数的数据

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

            由于字段含义中说："退货订单的代码以字母'c'开头"，因此怀疑是不是所有销量为负的订单是所谓的退货订单

            且目测所有的订单号码都是以C开头，因此提取同时满足这两个条件的数据

In [14]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

            由于在满足以上两种条件下仍输出了结果，因此认为我们上述的猜想并非正确

In [15]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

            根据输出结果猜想得到验证，当Quantity变量为负数的时候，观察值满足以下条件之一： 
                        1.InvoiceNo以C开头，表示订单被取消 
                        2.UnitPrice为0，说明单价为0英镑 这些交易数据均不是有效成交数据 不贡献销售不在后续的分析范围内
                        
            因此我们将在数据清理步骤，将Quantity为负数的观察值删除

In [16]:
original_data[(original_data["UnitPrice"] < 0)]    

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

            发现 "Description" 列中显示 "Adjust bad debt"，即坏账调整 不属于商品交易数据 因此也在数据清洗步骤中也将其删除

## 清洗数据

            根据前面评估部分得到的结论 我们需要进行的数据清理包括：

                        1.把InvoiceDate 变量的数据类型转换为日期时间
                        2.把CustomerID变量的数据类型转换为字符串
                        3.把description变量缺失的观察值删除
                        4.把country变量值"USA"替换为"United States"
                        5.把country变量值"UK"、"U.K."替换为”United Kingdom"
                        6.把Quantity变量值为负数的观察值删除
                        7.把Unitprice变量值为负数的观察值删除

            为了区分经过清理的数据和原始数据，我们创建新的变量cleand_data
            
            让它为original_data复制出的副本 我们之后的清理步骤将被运用在cleand_data上

In [17]:
cleaned_data = original_data.copy()    # 复制副本
cleaned_data.head()                    # 提取副本的前五行

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

            1.把InvoiceDate 变量的数据类型转换为日期时间

In [18]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])    # 调用.to_datetime()转换日期格式
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

            2.把CustomerID变量的数据类型转换为字符串

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)    # 调用.astype(str) 转化为object字符串
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

                    把CustomerID结尾的.0删除

In [20]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)    # .str.slice(保留的起点, 不保留的终点)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

            3.把Description变量缺失的观察值删除 并查看删除后该列空缺值个数和

In [21]:
cleaned_data.dropna(subset=["Description"], inplace=True)    # .dropna(subset=["Description"])保留 "Description" 无缺失值的部分
cleaned_data["Description"].isnull().sum()                   # 求 "Description" 有缺失值的和----bool被视为1or0

np.int64(0)

            4.country变量值"USA"替换为"United States",并检查替换后"USA"变量值个数

In [22]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})    # 调用.replace({"老值 ": "新值 "}) 
len(cleaned_data[cleaned_data["Country"]== "USA"])

0

            5.把country变量值"UK"、"U.K."替换为"United Kingdom" ,并检查替换后"UK"，"U.K."数量

In [23]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"} )

print(len(cleaned_data[cleaned_data["Country"]== "UK"]))
print(len(cleaned_data[cleaned_data["Country"]== "U.K"]))

0
0


            6.把Quantity变量值为负数的观察值删除并检查替换后Quantity变量值为负数的个数

In [24]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]    # 反向操作：筛选出 "UnitPrice"] >= 0 的数据，并进行赋值
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

            7.把UnitPrice变量值为负数的观察值删除并检查替换后UnitPrice变量值为负数的个数

            7.把UnitPrice变量值为负数的观察值删除并检查替换后UnitPrice变量值为负数的个数

In [25]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]    # 同理进行反向操作：因为没学过删除特定行的命令
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

In [26]:
cleaned_data.to_csv("e_commerce_cleanned.csv", index=False)    # 在原始csv同一文件夹下保存 "e_commerce_cleanned.csv"
                                                                # 并调用 index=False 不保存原始位置索引

In [27]:
pd.read_csv("/Users/warma/Desktop/python/实践项目：英国电商销售数据/e_commerce_cleanned.csv").head() 

/var/folders/nm/k6xd7m5j5_n5tkt_d9xcbflh0000gn/T/ipykernel_81328/1913380635.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("/Users/warma/Desktop/python/实践项目：英国电商销售数据/e_commerce_cleanned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom